### ISSUE TITLE
hostlist: handle IPv6 brackets

### ISSUE THREAD

**Issue by chu11** [flux-framework/flux-core#3282]
IPv6 supports brackets to differentiate between the colon for the IPv6 address and a colon for the port.  For example, `[::1]:5`, where `::1` is the address and `5` is the port number.

This brackets notation can confuse the hostlist library.  I added special case handling for this case in FreeIPMI ( See: https://git.savannah.gnu.org/cgit/freeipmi.git/tree/common/miscutil/fi_hostlist.c) but general support should be added to the hostlist library.

**Comment by grondo** [comment]
Could you add some test input here? I'm missing something obvious because I don't see how IPv6 addresses apply to the hostlist library?

**Comment by grondo** [comment]
We may also need to update [RFC 29](https://flux-framework.readthedocs.io/projects/flux-rfc/en/latest/spec_29.html), which states:

> The prefix and suffix components of a hostlist expression SHALL consist of any printable, non-whitespace ASCII character besides the special characters including square brackets: “[ ]”, and comma: “,”.

Therefore, currently an IPV6 address of the form above is not a valid Hostlist Expression. Not sure how to relax the RFC to allow this exception exactly... It may be simpler to leave it to the program calling hostlist functions to give "clean" input, e.g. when IP addrs are expected in input they can be pre-processed to remove invalid brackets for IPv6?

**Comment by chu11** [comment]
wanted to link #949, just so the IPv6 issues are connected

**Comment by chu11** [comment]
> Therefore, currently an IPV6 address of the form above is not a valid Hostlist Expression. Not sure how to relax the RFC to allow this exception exactly... It may be simpler to leave it to the program calling hostlist functions to give "clean" input, e.g. when IP addrs are expected in input they can be pre-processed to remove invalid brackets for IPv6?

That's a fair point.  I think in FreeIPMI I used the hostlist data structure as the primary "store the hosts the user inputted" data structure, and the hosts stored in that data structure were ultimately passed to functions that would check the host/IP for a port.

So it would have been a major effort to refactor with a new data structure that would pre & post process the hosts.  So some workarounds for this case were added to avoid that major refactor.

But for anyone using the library starting today, they would hopefully know up front to deal with this corner case.

**Comment by chu11** [comment]
> Could you add some test input here? I'm missing something obvious because I don't see how IPv6 addresses apply to the hostlist library.

Oh, missed this earlier comment.  In the FreeIPMI case, the hostlist data structure was the primary "store hosts the user inputted" data structure.  So hypothetically the user could do the following:

`ipmi-sensors -h myhost[1-4],192.168.0.1:123`

this is fine, if for some reason no hostname was assigned to that particular IP address.  But if the user inputted an IPV6 address

`ipmi-sensors -h myhost[1-4],[::1]:123`

the hostlist library would bork on this.

Although the use case seems rare to me, the fix request / initial patch came from Ubuntu originally.  So some set of users out there hit this problem.

Ultimately, in the current broker `bootstrap` code, we have code that works something like this:

```
for all "host" specified in bootstrap.toml:
     if "host" a range of hosts == "yes":
          add each "host" to json array
     else if == "no":
          add "host" to json array
     else error:
          handle error
```

basically if the user inputs a IPv6 address, how will this code change?  B/c the "is "host" a range of hosts" will presumably return an error indicating the input is invalid.

**Comment by grondo** [comment]
> basically if the user inputs a IPv6 address w/ a bracket

Maybe we can tackle this one later. I've never seen anyone configure a cluster with IP addresses instead of hostnames. Since Slurm uses a version of the old hostlist code, I'm assuming nobody has ever tried this with Slurm (though maybe we should peek at their modifications to the code to see if this use case has ever come up)

**Comment by garlick** [comment]
In case it's not clear, the hostnames used in the [boostrap] config must match the output of `hostname`, so I'm not sure that same issue applies in that case.

**Comment by chu11** [comment]
By all means it is not a high priority, just wanted to document this somewhere as I recall the issue popping up in FreeIPMI.   To be honest, I'm not entirely sure why anyone would configure with an IP instead of a hostname.  I suspect it may be VM / cloud kinds of scenarios where its easier to just grab a list of IPs you've been allocated than to assign hostnames to the IPs.

**Comment by grondo** [comment]
That's actually a good point, configuring an instance in the cloud may be a good use case to keep in mind!

(Perhaps we can have another "bootstrap" method for this case, though, treated more like starting an instance under a foreign resource manager)

**Comment by chu11** [comment]
FWIW, the way I handled this in FreeIPMI was to create a wrapper layer above `hostlist.[ch]`.  In it, all hostlist create/push/append/etc. functions would effectively detect the special case of `[IPV6address]:port` on input, then do some string substitutions that made the hostname illegal but processable by hostlist it (i.e. I `[::1]:80` -> `**L**::1**R**:80`).  Then on anything that decodes the hostlist, detect the special substitution strings and do the opposite conversion before returning the string to the user.

I did not handle something like `[1::1::1::[0-4]]:port`, which would have been harder to fix.
